[3.x] Promote object validity checks to release builds

RandomShaper opened the pull request:
Extra:
- Optimized the debug-only check about why the object is null to determine if it's because it has been deleted (the RC is enough; no need to check the ObjectDB).
- Because of the previous point. the debugger being attached is not required anymore for giving the "Object was deleted" error; from now, it only matters that it's a debug build.
- `is_instance_valid()` is now trustworthy. It will return `true` if, and only if, the last object assigned to a `Variant` is still alive (and not if a new object happened to be created at the same memory address of the old one).
- Replacements of `instance_validate()` are used where possible `Variant::is_invalid_object()` is introduced to help with that. (GDScript's `is_instance_valid()` is good.)

I've smoke-tested this on the Windows editor and on an Android release build of my game. No crashes. No apparent regressions.

Some of all this is still needed in 4.0. That work will happen separately.

Implements https://github.com/godotengine/godot-proposals/issues/1589.

**UPDATE:** ~~To be honest, I'm not very fond of the changes to `Tween` so it can robuslty check the validity of the objects involved. I'd like opinions.~~ Fixed.

akien-mga commented:
> UPDATE: To be honest, I'm not very fond of the changes to Tween so it can robuslty check the validity of the objects involved. I'd like opinions.

The main change seems to be the change of arguments from `Object *` to `Variant`. Why is it needed exactly? Since it seems the first thing you do anyway is to cast them to `Object *`. Does the validation happen only when casting Variants?

RandomShaper committed:
Promote object validity checks to release builds

Extra:
- Optimized the debug-only check about why the object is null to determine if it's because it has been deleted (the RC is enough; no need to check the ObjectDB).
- Because of the previous point. the debugger being attached is not required anymore for giving the "Object was deleted" error; from now, it only matters that it's a debug build.
- `is_instance_valid()` is now trustworthy. It will return `true` if, and only if, the last object assigned to a `Variant` is still alive (and not if a new object happened to be created at the same memory address of the old one).
- Replacements of `instance_validate()` are used where possible `Variant::is_invalid_object()` is introduced to help with that. (GDScript's `is_instance_valid()` is good.)

RandomShaper commented:
Conversion of `Variant` to `Object *` in C++ returns `null` if the object has been released (like the gone decay to null, but only for C++). Therefore, I thought that `Tween` needed to get `Variant`s so it could do the conversion to `Object *` itself. However, that was overdoing it, since the binding layer, so to speak, already does the conversion and `Tween` can just check for null.

The only change needed is to remove the calls to `ObjectDB::instance_validate()`, since they add no value over the null check. So, that's what I've done in my last push seconds ago.

akien-mga commented:
Thanks!

oeleo1 commented:
@RandomShaper Some commits deserve triple hearts and this is one of them! Thank you. Another was an off by one error in LocalVector fixed by @groud. Looking forward to the next 3.4 release.

RandomShaper commented:
@oeleo1, I really appreciate your words. I hope this is indeed as beneficial as we expect!

oeleo1 commented:
@RandomShaper Well, this is a game changer indeed. Automatic testing of release builds in 3.1 used to crash within the minute. 3.2 within 15 minutes, 3.4beta5 is still running after 6 hours. That's unseen unless it was a debug build. What a relief... simply fantastic. Bravo!
